File: #0648.replace-words.cpp
class Solution {
public:
    string replaceWords(vector<string>& dictionary, string sentence) {

        unordered_set<string> s;
        for (string &d : dictionary) {
            s.insert(d);
        }

        int size = sentence.size();
        string result;
        string cur;
        for (int i = 0, j; i <= size; i++) {
            if (i < size && !isspace(sentence[i])) {
                cur += sentence[i];
                continue;
            }
            
            string x;
            for (j = 0; j < cur.length(); j++) {
                x += cur[j];
                if (s.find(x) != s.end()) {
                    result += x + ' ';
                    break;
                }
            }

            if (j == cur.length()) {
                result += x + ' ';
            }
            cur.clear();
        }
        return result.substr(0, result.length() - 1);
    }
};

File: #0648.replace-words.js
/**
 * @param {string[]} dictionary
 * @param {string} sentence
 * @return {string}
 */
var replaceWords = function(dictionary, sentence) {
    dictionary.sort((a, b) => a.length - b.length);
    sentence = sentence.split(/\s+/);
    for (let i = 0; i < sentence.length; i++) {
        for (let d of dictionary) {
            if (sentence[i].startsWith(d)) {
                sentence[i] = d;
                break;
            }
        }
    }
    return sentence.join(' ');
};
